### ISSUE TITLE
Inconsistent read and write, which use the same predicate (WHERE clause)

### ISSUE THREAD

**Issue by JZuming** [pingcap/tidb#38150]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

**Setup the environment:**

```
tiup playground --db.binpath /path/to/latest/tidb-server &
```

**Test case**
```
mysql -h "127.0.0.1" -u root -P 4000 -D testdb < mysql_bk.sql
mysql -h "127.0.0.1" -u root -P 4000 -D testdb

mysql> select
  t_cp0sl.wkey as c0,
  t_cp0sl.pkey as c1,
  t_cp0sl.c_1_kgbc as c2,
  t_cp0sl.c_dw8ly as c3
from
  t_cp0sl
where t_cp0sl.c_1_kgbc not in (
  select
      subq_0.c0 as c0
    from
      (select
            ref_0.c_p5i5kc as c0,
            ref_0.wkey as c1
          from
            t_d0c_g as ref_0
          window w_9grx2d as ( partition by ref_0.pkey order by ref_0.c_eephud desc)) as subq_0
    where 77 not in (
      select
          1 as c0
        from
          t_d0c_g as ref_1));

mysql> update t_cp0sl set
  wkey = 59
where t_cp0sl.c_1_kgbc not in (
  select
      subq_0.c0 as c0
    from
      (select
            ref_0.c_p5i5kc as c0,
            ref_0.wkey as c1
          from
            t_d0c_g as ref_0
          window w_9grx2d as ( partition by ref_0.pkey order by ref_0.c_eephud desc)) as subq_0
    where 77 not in (
      select
          1 as c0
        from
          t_d0c_g as ref_1));
 
mysql> select
  t_cp0sl.wkey as c0,
  t_cp0sl.pkey as c1,
  t_cp0sl.c_1_kgbc as c2,
  t_cp0sl.c_dw8ly as c3
from
  t_cp0sl
where wkey = 59;
```

**mysql_bk.sql:** [mysql_bk.txt](https://github.com/pingcap/tidb/files/9640560/mysql_bk.txt)

### 2. What did you expect to see? (Required)

The first SELECT and the UPDATE should handle the same rows because they use the same predicate (WHERE clause) and execute adjacently. 

### 3. What did you see instead (Required)

The first SELECT outputs empty while the UPDATE changes 3 rows. To make sure what the UPDATE changes, I use the second SELECT, and it outputs the changed 3 rows.

**Results of the first SELECT statement in Test case**
```
Empty set (0.00 sec)
```

**Results of the UPDATE statement in Test case**
```
Query OK, 3 rows affected (0.01 sec)
Rows matched: 3  Changed: 3  Warnings: 0
```

**Results of the second SELECT statement in Test case**
```
+------+-------+------+--------+
| c0   | c1    | c2   | c3     |
+------+-------+------+--------+
|   59 | 15000 | NULL |  75.37 |
|   59 | 16000 | NULL | 100.57 |
|   59 | 17000 | NULL |  34.89 |
+------+-------+------+--------+
3 rows in set (0.01 sec)
```

Based on the analysis, I think it might be a bug triggering inconsistent read (the first SELECT) and write (the UPDATE).

### 4. What is your TiDB version? (Required)

```
Release Version: v6.3.0-20220913
Edition: Community
Git Commit Hash: 95e09ba33c0bdb629c4d4a7a31a2e4bc1212ad2a
Git Branch: HEAD
UTC Build Time: 2022-09-24 12:55:50
GoVersion: go1.19.1
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
```

**Comment by ljluestc** [comment]
```
-- Initiate a transaction
BEGIN;

-- Perform the SELECT operation
SELECT
  t_cp0sl.wkey as c0,
  t_cp0sl.pkey as c1,
  t_cp0sl.c_1_kgbc as c2,
  t_cp0sl.c_dw8ly as c3
FROM
  t_cp0sl
WHERE t_cp0sl.c_1_kgbc NOT IN (
  SELECT
    subq_0.c0 as c0
  FROM
    (SELECT
      ref_0.c_p5i5kc as c0,
      ref_0.wkey as c1
    FROM
      t_d0c_g as ref_0
    WINDOW w_9grx2d AS (PARTITION BY ref_0.pkey ORDER BY ref_0.c_eephud DESC)) AS subq_0
  WHERE 77 NOT IN (
    SELECT
      1 AS c0
    FROM
      t_d0c_g AS ref_1
  )
);

-- Perform the UPDATE operation
UPDATE t_cp0sl SET
  wkey = 59
WHERE t_cp0sl.c_1_kgbc NOT IN (
  SELECT
    subq_0.c0 as c0
  FROM
    (SELECT
      ref_0.c_p5i5kc as c0,
      ref_0.wkey as c1
    FROM
      t_d0c_g as ref_0
    WINDOW w_9grx2d AS (PARTITION BY ref_0.pkey ORDER BY ref_0.c_eephud DESC)) AS subq_0
  WHERE 77 NOT IN (
    SELECT
      1 AS c0
    FROM
      t_d0c_g AS ref_1
  )
);

-- Commit the transaction
COMMIT;

-- Validate the changes
SELECT
  t_cp0sl.wkey as c0,
  t_cp0sl.pkey as c1,
  t_cp0sl.c_1_kgbc as c2,
  t_cp0sl.c_dw8ly as c3
FROM
  t_cp0sl
WHERE wkey = 59;

```

**Comment by Defined2014** [comment]
It's been fixed in v7.5.0 and later, you could update your TiDB to fix this problem.
